### PR TITLE
Index block predecessors for CFG profiling

### LINKED ISSUE #238: jit: Add predecessor-indexed CFG profiling infrastructure
## Goal

Improve control-flow profiling and block-cache maintenance without committing to a graph-based SSA IR or whole-function lifting.

## Context

The emulator already records learned taken and untaken block edges, tracks indirect-target history, and uses block chaining to bypass the dispatcher. Current work also includes packed interpreter traces and trace-local fusion. The missing capability is efficient reverse-edge tracking: when a cached block is evicted or invalidated, its incoming links must currently be found by scanning translated blocks.

A single `prev` pointer is not sufficient because a CFG node can have multiple predecessors, including indirect control-flow targets.

## Proposed scope

- Introduce a bounded, ownership-safe predecessor index for translated-block entry edges.
- Keep it consistent when direct, conditional, and learned indirect edges are installed, replaced, evicted, or invalidated.
- Use it to unlink incoming edges without a full block-list traversal.
- Expose CFG-edge and hot-path profiling data through the existing profiler/tools.
- Add focused tests for edge installation, multiple predecessors, cache eviction, invalidation, and stale-edge prevention.

## Non-goals

- No whole-function lifting based on ELF metadata.
- No new graph/SSA/use-def IR in this issue.
- No cross-function optimization or replacement of the existing Tier-1 or LLVM Tier-2 pipelines.

## Success criteria

- Correctness is preserved under cache replacement, address-space changes, and indirect branches.
- The implementation removes or materially reduces global scans during block unlinking.
- Profiling identifies useful hot-edge/trace candidates.
- Benchmark results show neutral or positive steady-state performance and acceptable memory overhead.

## Follow-up decision

Consider a trace-local or graph-IR prototype only if this infrastructure and benchmark data demonstrate an optimization opportunity that the existing packed IR, block chaining, fusion, and LLVM Tier-2 compiler cannot address.

### PR DESCRIPTION
**Pull request by jserv**
Evicting a cached block had to find the blocks pointing at it, and the only way to do that was to walk every translated block and compare its tail's two branch slots, with a TODO in the code admitting as much. Each block now carries the two outgoing edges it can ever have, linked into the destination's incoming list, so either endpoint unlinks in constant time. The patch site is derived from the source block's current tail rather than pinned at link time, which keeps an edge valid when macro-op fusion shortens a block that is already chained. Edge ownership stays on the emulator thread: the compiler thread signals an evicted block through should_free and no longer reaps edges itself, which is what lets the chaining step run without taking cache_lock on every block transition. The one remaining crossing, where t2c_trace_ebb reads a chain pointer the emulator may be installing, is published and read with relaxed atomics. The -p report gains the predecessor count, the tier-1 re-entry count and the observed indirect history, with a header saying what each column does and does not count.

Measuring this turned up a bug that predates the branch. muldivmod picks the x86-64 divide from the opcode class and picked the unsigned form for every caller, so DIV and REM returned the unsigned quotient whenever an operand was negative, and INT_MIN / -1 raised #DE and killed the emulator with SIGFPE. It is fixed here in its own commit, along with the reason nothing caught it: no test applied a JIT-compiled arithmetic instruction to an awkward operand and checked the answer. There is now a differential sweep that applies every RV32IM arithmetic form to each ordered pair from a table of awkward operands, plus loads, stores and branches, and compares the result against the interpreter.

Verified on jit_defconfig, system_jit_defconfig, system_defconfig, a tier-1-only JIT build and an interpreter build: all build warning-free and pass make check and make tests. The guest regressions pass on each configuration that can run them. UBSAN is clean across the regressions, the benchmarks and both suites on the JIT and interpreter builds. Valgrind reports the same leak profile as master and no new errors on an eviction-heavy workload. Merging onto current master builds and passes tests.

Two things are deliberately left out. The RISC-V conformance suite was not run, because riscof and a bare-metal toolchain are unavailable here, so the differential sweep stands in for it rather than replacing it; it is worth running over the muldivmod change before merging. And try_lazy_fusion still mutates and frees a live block's IR list on the cache-hit path without cache_lock, while the compiler thread may be walking the same list. That predates this work and is reachable in system_jit_defconfig, but fixing it is a design decision rather than part of this change.

Closes #238

---
## Summary by cubic
Indexes CFG predecessor edges so cache eviction unlinks chains in constant time instead of scanning translated blocks, and removes `cache_lock` from normal chaining. It also fixes tier-1 JIT correctness bugs found during profiling, including signed `DIV`/`REM`, memory addressing, register aliases, and result normalization; closes #238.

**Refactors**

- Source-owned edges link into destination incoming lists, with patch sites derived from the source tail so macro-op fusion preserves chains.
- The emulator thread performs deferred edge cleanup, while the compiler reads chain pointers with relaxed atomics.
- Cache lookups return frequency in one pass, loop tracking uses generation stamps, and `-p` labels its expanded profiling columns.
- Indirect-target history retains warm-up counts and can specialize two hot targets; this stays disabled for tiered builds until chained code participates in tiering.
- `-p` derives report paths from the executable directory, and profiling and test artifacts are ignored.

**Bug Fixes**

- Signed division uses 64-bit host `IDIV`, so negative operands produce correct RV32 results and `INT_MIN / -1` no longer raises `SIGFPE`.
- New regressions cover edge ownership and eviction, loop and branch history, JIT ALU and memory behavior, indirect targets, and an interpreter differential sweep.
- JIT and interpreter builds pass the available checks, UBSAN, and Valgrind; RISCOF was not run because its toolchain is unavailable.
- The pre-existing `try_lazy_fusion` live-IR race remains out of scope.

<sup>Written for commit d46373953aa4e8e0c792eef15fda4ecf5b6ae9fa. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/sysprog21/rv32emu/pull/780?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Review by jserv — COMMENTED** [review]
# Benchmarks

<details>

| Benchmark suite | Current: d46373953aa4e8e0c792eef15fda4ecf5b6ae9fa | Previous: 3531d82d6f0590d0f504ec83fc3ea9074d5cc203 | Ratio |
|-|-|-|-|
| `Dhrystone` | `2189.75` DMIPS | `2118.5` DMIPS | `0.97` |
| `CoreMark` | `1742.075` iterations/sec | `1745.557` iterations/sec | `1.00` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).

File: src/jit.c [hunk 1]
```diff
@@ -1470,14 +1651,27 @@ static void muldivmod(struct jit_state *state,
...
+        }
     }
 
-    if (is64)
+    /* A signed divide runs 64-bit over the sign-extended operands. That keeps
+     * INT_MIN / -1 representable, so the instruction cannot raise #DE, and its
+     * low half is the result RV32 defines. The overflow fixup below then
+     * agrees with it rather than correcting it.
+     */
+    if (is64 || sign)
```
[the remark this answers is not in the document]
Comment (jserv) [reply src/jit.c:1681]: Confirmed and fixed in 3531d82. You are right that the 64-bit IDIV leaves a negative result sign-extended and that the writeback is a 64-bit move, so the mapped register stopped holding the zero-extended word the rest of the generator assumes.

I reproduced it before changing anything: dividing the stack pointer (0xFFFFE000, negative as int32) and loading through the quotient gives rc=1 under the JIT and rc=0 under the interpreter. The case only fires when nothing spills in between — my first attempt put a `bne` between the divide and the load, which forced a `store_back` and reloaded the base zero-extended, hiding it.

The fix narrows the result after the writeback, and unconditionally, so it also covers `dst == RAX` where no move happens. `tests/jit-signed-div.S` now carries the reproducer.

File: src/decode.h [hunk 2]
```diff
@@ -995,19 +995,28 @@ typedef struct {
...
+    const uint32_t idx = (pc >> 2) & (HISTORY_SIZE - 1);
+    bool match = bt->PC[idx] == pc;
+#if RV32_HAS(SYSTEM)
+    match = match && bt->satp[idx] == satp;
+#endif
+    if (match) {
+        if (bt->times[idx] != UINT32_MAX)
+            bt->times[idx]++;
+    } else {
+        bt->PC[idx] = pc;
```
[the remark this answers is not in the document]
Comment (jserv) [reply src/decode.h:1014]: Fixed in 3531d82, and the reasoning is a little stronger than stated: the table memsets `PC` to all ones rather than to zero, so `!bt->PC[idx]` never rejected an unobserved slot in the first place — `times` was already doing that work. The only thing the PC test could reject was a genuine hot target at guest address 0.

So the check is now just `bt->times[idx] < IN_JUMP_THRESHOLD`, which is both correct and shorter. Low impact in practice, since an indirect jump to address 0 is a null function pointer the guest would fault on, but there is no reason to keep a sentinel that cannot do its job.

File: mk/tests.mk [hunk 3]
```diff
@@ -91,9 +105,45 @@ $(foreach e,$(CHECK_ELF_FILES),$(eval $(call make-check-target,$(e))))
 GUEST_CFLAGS := -march=rv32i -mabi=ilp32
 GUEST_CHECKS := fused-misalign syscall-zero-write
 GUEST_CHECK_TARGETS :=
+GUEST_ASM_CHECK_TARGETS :=
 ifneq ($(CROSS_COMPILE),)
 ifneq ($(CONFIG_RV32E),y)
 ifneq ($(filter check,$(MAKECMDGOALS)),)
+# Freestanding regressions do not require the rv32 libc multilib.
+GUEST_ASM_WORKS := $(shell printf '' | $(CROSS_COMPILE)gcc $(GUEST_CFLAGS) \
```
[the remark this answers is not in the document]
Comment (jserv) [reply mk/tests.mk]: Fixed in 3531d82. Each ISA string is now probed separately — rv32i, rv32im, rv32ic and rv32imc — and each check is gated on the one it actually assembles with, so a toolchain that handles the base ISA but not the M or C variants skips those checks instead of failing the build.

Worth noting these are all `-nostdlib -static`, so no multilib is involved and only the assembler and linker need to accept the string, which makes the failure unlikely rather than impossible. The probe is cheap enough that unlikely is not a good reason to leave it unguarded.

File: mk/tests.mk [hunk 4]
```diff
@@ -91,9 +105,45 @@ $(foreach e,$(CHECK_ELF_FILES),$(eval $(call make-check-target,$(e))))
...
+# These run user-mode guest ELFs, which a system-mode emulator can only load
+# with the ELF loader selected; otherwise it expects a kernel image.
+RUN_USER_ELF :=
+ifneq ($(CONFIG_SYSTEM),y)
+RUN_USER_ELF := y
+else ifeq ($(CONFIG_ELF_LOADER),y)
+RUN_USER_ELF := y
+endif
+
+ifeq ($(GUEST_ASM_WORKS)$(RUN_USER_ELF),yy)
```
[the remark this answers is not in the document]
Comment (jserv) [reply mk/tests.mk:133]: Fixed in 3531d82. You are right that gating only the new checks made things worse rather than better — a system build still ran `check-hello` and the C guest checks and failed on them, which is exactly the inconsistency I introduced.

`RUN_USER_ELF` is now hoisted and gates every check that runs a guest ELF: `check-hello`, the C guest checks, `CHECK_ELF_FILES` and `check-rvv-smoke`, alongside the assembly ones. `check-trace-match` builds and runs a host program, so it stays ungated. With a cross toolchain present, `make -n check` now schedules zero ELF checks on system_defconfig and system_jit_defconfig, and the usual set on the user-mode configs.

File: tests/jit-alu-alias.S [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    li t3, 0x80000000
+    bne t0, t3, fail
+    li t3, 1
+    bne t1, t3, fail
+    .endm
+
+_start:
+    li s0, 20000
+loop:
+    aliases add, 0x9d0369ce, 0x13579bde
```
[the remark this answers is not in the document]
Comment (jserv) [reply tests/jit-alu-alias.S:54]: Added in 3531d82. The point lands because this PR actually changes both of those paths: `GEN_SHIFT_REG` lost a redundant `and` mask and had its move order changed, and `GEN_ALU_REG` gained a commutative shortcut that skips the temporary. Aliasing coverage for them is therefore about this diff, not hypothetical.

`aliases` now also runs sll, srl, sra, slt and sltu. The expected values differ from both operands in every case, so a stale source changes the observed result rather than coinciding with it.

One thing the new differential sweep does not cover, which is why this matters: it always writes to a distinct destination, so it exercises the operand space but never destination aliasing. The two tests are complementary.

File: .gitignore [hunk 6]
```diff
@@ -24,8 +29,16 @@ build/linux-x86-softfp/
...
+build/.effective-config
+build/.stamp-prebuilt
 build/bin2c
 build/minimal.dtb
+build/trace-match
+build/block-eviction
+build/jit-alu-alias
+build/jit-identity-normalize
+build/jit-indirect-targets
+build/jit-memory-address*
```
[the remark this answers is not in the document]
Comment (jserv) [reply .gitignore:41]: Fixed in 3531d82. Added `build/jit-signed-div` and `build/jit-interp-diff*` alongside the existing guest binary entries, plus `build/*/test-*` to cover the per-suite test binaries generally rather than adding a line per suite.

Context on why they were unignored: the per-suite directory entries were deliberately removed earlier in review, which left those binaries exposed. The single pattern covers the existing suites and any future one without reintroducing a line per test.

File: mk/tests.mk [hunk 7]
```diff
@@ -91,15 +105,65 @@ $(foreach e,$(CHECK_ELF_FILES),$(eval $(call make-check-target,$(e))))
...
 # An auto-detected toolchain may lack an rv32i/ilp32 multilib; probe it so such
 # a toolchain skips these programs instead of failing make check.
 GUEST_CC_WORKS := $(shell printf 'int main(void) { return 0; }\n' | \
 	$(CROSS_COMPILE)gcc $(GUEST_CFLAGS) -x c -o /dev/null - >/dev/null 2>&1 && \
 	echo y)
-ifeq ($(GUEST_CC_WORKS),y)
+ifeq ($(GUEST_CC_WORKS)$(RUN_USER_ELF),yy)
 GUEST_CHECK_TARGETS := $(addprefix check-,$(GUEST_CHECKS))
 else
 $(info Skipping guest regression programs: $(CROSS_COMPILE)gcc cannot build $(GUEST_CFLAGS) programs)
```
[the remark this answers is not in the document]
Comment (jserv) [reply mk/tests.mk]: Fixed. You are right, and I had the misleading output in front of me earlier without registering it: on system_jit_defconfig both lines printed, so the compiler got blamed for a skip that the ELF loader condition caused.

The two conditions are tested separately now. With a working cross toolchain, jit_defconfig schedules the guest C checks and prints nothing, while system_jit_defconfig prints only the user-ELF line and schedules none. The compiler message now appears only when the compiler is actually the problem.

This landed as part of the rebase onto current master, which also folded the review-response commits into the ones they were editing, so the branch is now six commits rather than eight.
